dali/kernels/kernel_manager.h

```diff
@@ -0,0 +1,237 @@
+
+  template <typename Kernel, typename... Args>
+  Kernel &create_or_get(Args&&... args) {
+    void (*deleter)(void *) = delete_kernel<Kernel>;
+    if (!instance || instance.get_deleter() != deleter) {
+      instance.reset();
+      instance = { new Kernel{std::forward<Args>(args)...}, deleter };
+    }
+    return *static_cast<Kernel*>(instance.get());
+  }
+
+  template <typename Kernel, typename... Args>
+  Kernel &get() {
+    void (*deleter)(void *) = delete_kernel<Kernel>;
+    if (!instance)
+      throw std::logic_error("The kernel instance is null");
+    if (instance.get_deleter() != deleter)
+      throw std::logic_error("The kernel instance is of different type than requested");
+
+    return *static_cast<Kernel*>(instance.get());
+  }
+
+  template <typename Kernel>
+  static void delete_kernel(void *ptr) {
+    delete static_cast<Kernel*>(ptr);
+  }
+
+  explicit operator bool() const noexcept { return static_cast<bool>(instance); }
+};
+
+/// @brief Manages multiple instances of run-time typed kernels
+///
+/// KernelManager provides type erasure for kernels whose type is selected at
+/// run-time. Kernel manager also carries out mundane tasks of keeping
+/// ScratchpadAllocators and reserving memory according to requirements returned
+/// by kernel's Setup method.
+///
+/// A scratchpad allocator is created per-thread with thread indexing supported
+/// explicitly by the caller.
+class DLL_PUBLIC KernelManager {
+ public:
+  using ScratchSizes = std::array<size_t, ScratchpadAllocator::NumAllocTypes>;
+
+  /// @brief Creates `num_threads` scratcapads and `num_instances` slots for kernels
+  ///
+  /// @param num_threads -    number of threads that can concurrently use the kernels in the
+  ///                         manager, assuming that each threads uses its unique
+  ///                         zero-based inde
+  /// @param num_instances -  number of Kernel instances to be created; typically corresponds
+  ///                         to number of samples (for per-sample kernels) or minibatches
+  void Initialize(size_t num_threads, size_t num_instances);
+
+  /// @brief Creates `num_threads` scratcapads and `num_instances` kernels of type Kernel
+  ///        constructed with `args...`.
+  ///
+  /// @param num_threads   -  number of threads that can concurrently use the kernels in the
+  ///                         manager, assuming that each threads uses its unique
+  ///                         zero-based inde
+  /// @param num_instances -  number of Kernel instances to be created; typically corresponds
+  ///                         to number of samples (for per-sample kernels) or minibatches
+  /// @param args           - arguments passed to Kernel's constructor upon creation.
+  /// @tparam Kernel        - type of the kernel to be created
+  template <typename Kernel, typename... Args>
+  void Initialize(size_t num_threads, size_t num_instances, const Args&... args) {
```

Comment:
Hmm, the downside of this is that we cannot have an initialize that does default-init, because the overload will only create the slots.
Maybe call the empty `Initialize` as something like `Allocate` or `MakeSlots`.

Reply:
We can.
Look at ResizeBase.